mock/tests/testmock.py

```diff
@@ -1433,6 +1435,32 @@ def f(a, b, c, d=None): pass
             mock.assert_has_calls(calls[:-1])
         mock.assert_has_calls(calls[:-1], any_order=True)
 
+    def test_assert_has_calls_not_matching_spec_error(self):
+        def f(x=None): pass
+
+        mock = Mock(spec=f)
+        mock(1)
+
+        with self.assertRaisesRegex(
+                AssertionError,
+                '^{}$'.format(
+                    re.escape('Calls not found.\n'
+                              'Expected: [call()]\n'
+                              'Actual: [call(1)]'))) as cm:
+            mock.assert_has_calls([call()])
+        self.assertIsNone(cm.exception.__cause__)
+
+
+        with self.assertRaisesRegex(
+                AssertionError,
+                '^{}$'.format(
+                    re.escape(
+                        'Error processing expected calls.\n'
+                        "Errors: [None, TypeError('too many positional arguments')]\n"
```

Comment:
The repr of TypeError differs in Python 3.6 (this patch wasn't backported to 3.6, so it doesn't account for that). In 3.6, there's a trailing comma after the single argument.

```none
$ python3.6
>>> TypeError('foo')
TypeError('foo',)

$ python3.7
>>> TypeError('foo')
TypeError('foo')
```

Presumably this can become:
```python
'^{}{}{}$'.format(
    re.escape(
      'Error processing expected calls.\n'
      "Errors: [None, TypeError('too many positional arguments'",
    ',?',
    re.escape(')]\n'))
```

Or something like that.

Reply:
Ah, great catch! Thanks, the workaround I've gone for is uglier, but hopefully will have future merges working better.